File: frontend/src/lib/media.ts
/**
 * What counts as media, for both halves of the product.
 *
 * The console writes these values into an article document and the public feed
 * renders them, so the two must agree about the shape of an image `src` and
 * about what a YouTube video id is. This module is where they agree — and it
 * imports nothing, which is the point: the reader's bundle can use it without
 * pulling in TipTap, and `lib/api` stays free of the editor for the same
 * reason it stays free of the router.
 *
 * The server has the last word on all of this (`backend/app/services/media.py`)
 * and refuses a save that breaks it. The checks here are so that a reviewer
 * finds out at the moment they paste something, rather than at the moment
 * their autosave fails.
 */

/**
 * The only image `src` an article may hold: the path the upload endpoint
 * returned, which is a content hash under a two-character shard.
 *
 * Mirrors `MEDIA_SRC_RE` in `backend/app/services/media.py`. Anchored, and no
 * `.` or `/` inside, so neither a remote URL nor `/api/media/../../etc/passwd`
 * matches. Images are uploaded, never linked: a linked image breaks when the
 * other site reorganises, and tracks the reader until it does.
 */
const MEDIA_SRC_RE = /^\/api\/media\/[0-9a-f]{2}\/[0-9a-f]{62}\.(?:png|jpg|gif|webp)$/;

/** YouTube ids are 11 characters of an unpadded base64url alphabet. */
const YOUTUBE_ID_RE = /^[A-Za-z0-9_-]{11}$/;

/** Hosts a watch link can arrive from. Exact, so `evilyoutube.com` is not one. */
const YOUTUBE_HOSTS = new Set([
  "youtube.com",
  "m.youtube.com",
  "music.youtube.com",
  "youtube-nocookie.com",
  "youtu.be",
]);

/**
 * How a picture or a video sits in the prose.
 *
 * `left` and `right` float, and the prose wraps down the other side — Word's
 * "Square" wrap. `none` is a block with text above and below; when it is
 * narrower than the column it centres, because a narrow block hugging the left
 * margin reads as a mistake rather than a choice.
 *
 * There are no coordinates here on purpose. Media are siblings of the beat
 * paragraphs and are placed by being dragged between them, which is what keeps
 * `attrs.beat` — and therefore the derived feed card — meaningful.
 */
export type MediaAlign = "none" | "left" | "right";

const MEDIA_ALIGNMENTS: readonly MediaAlign[] = ["none", "left", "right"];

/**
 * Width is a percentage of the prose column, never pixels.
 *
 * The editor's column is around 850px and the published article's is 64ch, so
 * a percentage is the only unit under which a reviewer's choice means the same
 * thing in both. The floor keeps a picture from being dragged down to a speck
 * that no reader can make anything out of.
 */
export const MEDIA_MIN_WIDTH = 20;
export const MEDIA_MAX_WIDTH = 100;

/** True for a path this system's own media store produced. */
export function isStoredMedia(src: unknown): src is string {
  return typeof src === "string" && MEDIA_SRC_RE.test(src);
}

export function isMediaAlign(value: unknown): value is MediaAlign {
  return MEDIA_ALIGNMENTS.includes(value as MediaAlign);
}

/**
 * The classes that lay a media block out, from `styles/evidwell.css`.
 *
 * Here rather than in either styles module because the console and the feed
 * must produce the *same* classes — that identity is what makes "the editor
 * shows what publishes" true of layout. Two copies of this mapping would be
 * two things that could drift, and the symptom of drift is an article that
 * looks one way to the reviewer approving it and another to everyone else.
 *
 * Pair it with a `--ew-media-width` custom property for the width.
 */
export function mediaWrapClass(align: MediaAlign): string {
  if (align === "left") return "ew-media ew-media--left";
  if (align === "right") return "ew-media ew-media--right";
  return "ew-media";
}

/**
 * The alignment on a node, defaulting for anything else.
 *
 * Absent is the common case, not an error: every document written before media
 * could be laid out has no `align`, and those must keep rendering.
 */
export function mediaAlign(value: unknown): MediaAlign {
  return isMediaAlign(value) ? value : "none";
}

/**
 * A width in range, or full width.
 *
 * Rounded to whole percents — the resize handle produces fractions of a pixel
 * and storing `44.7391304347826` makes a diff between the original and the
 * edited document unreadable for the sake of a difference nobody can see.
 */
export function clampMediaWidth(value: unknown): number {
  const width = typeof value === "number" ? value : Number(value);
  if (!Number.isFinite(width)) return MEDIA_MAX_WIDTH;
  return Math.min(MEDIA_MAX_WIDTH, Math.max(MEDIA_MIN_WIDTH, Math.round(width)));
}

export function isYouTubeId(value: unknown): value is string {
  return typeof value === "string" && YOUTUBE_ID_RE.test(value);
}

/**
 * The video id inside whatever the reviewer pasted, or null.
 *
 * Accepts the four links YouTube itself hands out — `watch?v=`, `youtu.be/`,
 * `/embed/`, `/shorts/` — and a bare id, because someone will paste one. Only
 * the id is kept: a document stores data, and the renderer builds the embed
 * URL from it, so no pasted string ever reaches an iframe's `src` intact.
 */
export function youtubeIdFromUrl(input: string): string | null {
  const trimmed = input.trim();
  if (!trimmed) return null;
  if (isYouTubeId(trimmed)) return trimmed;

  let url: URL;
  try {
    // Bare `youtu.be/xyz` is a URL to a human and a relative path to `URL`.
    url = new URL(/^https?:\/\//i.test(trimmed) ? trimmed : `https://${trimmed}`);
  } catch {
    return null;
  }

  const host = url.hostname.replace(/^www\./, "");
  if (!YOUTUBE_HOSTS.has(host)) return null;

  const lastSegment = url.pathname.split("/").filter(Boolean).pop() ?? "";
  const candidate = url.searchParams.get("v") ?? lastSegment;
  return isYouTubeId(candidate) ? candidate : null;
}

/**
 * The still frame, for the console's block and the feed's click-to-play cover.
 *
 * `i.ytimg.com` serves thumbnails without the player, so showing one costs a
 * reader an image request rather than YouTube's whole JavaScript bundle and
 * the cookies that come with it.
 */
export function youtubeThumbnailUrl(videoId: string): string {
  return `https://i.ytimg.com/vi/${videoId}/hqdefault.jpg`;
}

/**
 * The watch page, for the console's "open it and check" control.
 *
 * A reviewer sizing a video down to a third of the column can no longer read
 * its title off the still, and "is this the right video" is the one question
 * the block has to be able to answer.
 */
export function youtubeWatchUrl(videoId: string): string {
  return `https://www.youtube.com/watch?v=${videoId}`;
}

/**
 * The player, built here from an id that has been checked against
 * {@link isYouTubeId} — never from a string anyone typed.
 *
 * `youtube-nocookie.com` because a reader arriving at a health article has not
 * asked to be tracked by Google, and this is the domain that honours that
 * until they press play.
 */
export function youtubeEmbedUrl(videoId: string, { autoplay = false } = {}): string {
  const params = new URLSearchParams({ rel: "0", modestbranding: "1" });
  if (autoplay) params.set("autoplay", "1");
  return `https://www.youtube-nocookie.com/embed/${videoId}?${params}`;
}

File: backend/app/services/media.py
"""Article media: what the store accepts, and what a document may cite.

Two rules, which are one rule seen from either end:

* **Bytes enter the store only if their content says they are an image.** The
  client's ``Content-Type`` and filename are read and discarded — both are
  attacker-controlled strings, and what this store holds is served straight
  back over HTTP. The first bytes of the file decide.
* **A document may reference only media this store wrote**, plus a YouTube
  video id. A remote ``https://`` image, a ``javascript:`` URL, a hand-typed
  iframe src: all refused, at save time and again at approve time.

SVG is deliberately absent from the allowlist. An SVG is a script-bearing
document; one served from our own origin and embedded in a published article is
stored XSS against every reader of that article. PNG, JPEG, GIF and WebP cannot
carry script, which is the whole reason the list is those four.

Storage is content-addressed — the key is the SHA-256 of the bytes. The same
image stored twice is one row, two writers can never collide, and no fragment
of a client-supplied string reaches the key.

**The bytes live in Postgres**, in ``media_objects``, keyed by that digest. They
were on local disk until they turned out to be the one piece of published state
that a database backup did not cover and a container did not carry — a run on
the host wrote files a containerised API then served as 404s, with the paths in
``articles.generated_imagery`` pointing at each one. One store, one backup, one
restore. See ``migrations/0004_media_objects.sql`` for why the table is keyed by
digest rather than hung off ``articles``.

The module is split so the part worth testing hardest needs no database:
``prepare_image()`` is pure — it sniffs, digests, and builds the path a document
will hold — and ``store_image()`` is that plus one INSERT. The seam is still
narrow enough that an S3 version is this module and nothing else (DESIGN.md §11
— deployment is deferred).

Orphan collection is deferred, as it was on disk: an image dropped from a draft
leaves its row behind. Rows are immutable and content-addressed, which makes a
sweep over ``articles.original_content``/``edited_content`` a job that can be
written correctly later — and an easier one now that it is a DELETE with no
filesystem to keep in step.
"""

from __future__ import annotations

import hashlib
import re
from dataclasses import dataclass
from typing import Any

from sqlalchemy import select
from sqlalchemy.dialects.postgresql import insert as pg_insert
from sqlalchemy.exc import IntegrityError
from sqlalchemy.ext.asyncio import AsyncSession

from app.domain.models import MediaObject
from app.services.tiptap import iter_nodes

#: Where stored media is served from. Under ``/api`` on purpose: the frontend
#: dev server already proxies that prefix and any deployment already routes it,
#: so uploaded images need no new rule in either place. Article documents hold
#: this path verbatim, so changing it is a rewrite of stored JSON.
MEDIA_URL_PREFIX = "/api/media"

#: Magic bytes → extension. Order matters only in that each prefix is checked
#: whole; none of the four is a prefix of another.
_SIGNATURES: tuple[tuple[bytes, str], ...] = (
    (b"\x89PNG\r\n\x1a\n", "png"),
    (b"\xff\xd8\xff", "jpg"),
    (b"GIF87a", "gif"),
    (b"GIF89a", "gif"),
)

#: Longest signature we need to see before deciding. WebP needs 12.
SNIFF_BYTES = 12

#: Extension → the type ``StaticFiles`` will serve the file as. Exists so the
#: upload response can report what the bytes *were*, rather than echoing back
#: the Content-Type the client claimed and which nothing checked.
CONTENT_TYPES: dict[str, str] = {
    "png": "image/png",
    "jpg": "image/jpeg",
    "gif": "image/gif",
    "webp": "image/webp",
}

#: The exact shape ``store_image`` produces: a two-character shard, the
#: remaining 62 hex characters of the digest, one of four extensions. Anchored,
#: with no ``.`` or ``/`` permitted inside, so ``/api/media/../../etc/passwd``
#: does not match and neither does a same-origin path pointing anywhere else.
MEDIA_SRC_RE = re.compile(
    rf"^{re.escape(MEDIA_URL_PREFIX)}/[0-9a-f]{{2}}/[0-9a-f]{{62}}\.(?:png|jpg|gif|webp)$"
)

#: YouTube video ids are exactly 11 characters of an unpadded base64url
#: alphabet. Documents store the **id**, never a URL: the renderer builds the
#: embed src from it, so there is no string a reviewer can supply that lands
#: unexamined in an iframe.
YOUTUBE_ID_RE = re.compile(r"^[A-Za-z0-9_-]{11}$")

#: How a media block sits in the prose. ``left`` and ``right`` float and the
#: text wraps down the other side; ``none`` is a block. Mirrors ``MediaAlign``
#: in ``frontend/src/lib/media.ts``.
#:
#: An enum rather than a CSS string is the whole point. The renderer turns one
#: of these three words into a class it already ships; nothing a reviewer
#: supplies becomes style on a published page.
MEDIA_ALIGNMENTS = frozenset({"none", "left", "right"})

#: Width is a percentage of the prose column, never pixels — the editor's
#: column and the article's differ, and a percentage is the only unit that
#: means the same in both. The floor stops a picture being shrunk to a speck.
MEDIA_MIN_WIDTH = 20
MEDIA_MAX_WIDTH = 100


class UnsupportedMediaError(ValueError):
    """Uploaded bytes are not one of the four accepted image formats."""


class UnsafeMediaError(ValueError):
    """A document references media this system did not store.

    Surfaces as HTTP 409 through ``ReviewError``. Never softened to a warning
    and never fixed by silently dropping the node — a reviewer whose image
    vanished without explanation has no way to tell that from a bug.
    """


def sniff_image(data: bytes) -> str | None:
    """The file extension the bytes actually are, or None.

    Returns the extension rather than a MIME type because the extension is what
    gets written to disk, and deriving it from anything the client said is the
    mistake this function exists to prevent.
    """
    for signature, extension in _SIGNATURES:
        if data.startswith(signature):
            return extension
    # WebP is a RIFF container; the format tag sits at offset 8.
    if data[:4] == b"RIFF" and data[8:12] == b"WEBP":
        return "webp"
    return None


@dataclass(frozen=True, slots=True)
class StoredImage:
    """An image now in the store. ``src`` is what the document holds."""

    src: str
    content_type: str
    size: int


@dataclass(frozen=True, slots=True)
class PreparedImage:
    """Everything derived from the bytes, before anything is written.

    Split out from ``store_image`` so the decisions worth testing hardest — is
    this really an image, what is it keyed by, what path will a document hold —
    are a pure function with no database in reach. The tests that pin them run
    on every ``pytest`` rather than only when Postgres is up, which matters:
    seven suites already skip silently without one.
    """

    digest: str
    extension: str
    data: bytes

    @property
    def src(self) -> str:
        """The path a document holds, and the one ``MEDIA_SRC_RE`` matches."""
        return f"{MEDIA_URL_PREFIX}/{self.digest[:2]}/{self.digest[2:]}.{self.extension}"

    @property
    def content_type(self) -> str:
        return CONTENT_TYPES[self.extension]

    def stored(self) -> StoredImage:
        return StoredImage(
            src=self.src, content_type=self.content_type, size=len(self.data)
        )


def prepare_image(data: bytes) -> PreparedImage:
    """Sniff, digest, and work out the path — touching nothing.

    Raises:
        UnsupportedMediaError: the bytes are not PNG, JPEG, GIF or WebP.
    """
    extension = sniff_image(data[:SNIFF_BYTES])
    if extension is None:
        raise UnsupportedMediaError(
            "not a PNG, JPEG, GIF or WebP image (the file's own contents were "
            "checked, not its name)"
        )
    return PreparedImage(
        digest=hashlib.sha256(data).hexdigest(), extension=extension, data=data
    )


async def store_image(data: bytes, *, session: AsyncSession) -> StoredImage:
    """Put image bytes in the store and return the path a document may hold.

    **Does not commit.** The caller owns the transaction — which for the
    pipeline means the orchestrator, whose rule that no stage commits is not
    relaxed for this one. The consequence is that ILLUSTRATE now writes rows
    rather than files, so its pictures land with the stage commit and vanish
    with the stage rollback. That is strictly better than the disk behaviour it
    replaces, where a failed run left its images behind as orphans.

    Storing the same image twice is a no-op, not a duplicate and not an error:
    the digest is the primary key, and ``ON CONFLICT DO NOTHING`` is exact here
    in a way it is not for ``SourceCache`` — one key, one inference clause,
    none of the partial-index ambiguity that made that upsert a resolve-first.

    The insert is wrapped in a ``SAVEPOINT`` for the same reason the source
    cache's retry is: the pipeline session is long-lived, and letting an
    IntegrityError poison it would discard everything the run has done so far
    over a decorative picture.

    Raises:
        UnsupportedMediaError: the bytes are not PNG, JPEG, GIF or WebP.
    """
    prepared = prepare_image(data)

    statement = (
        pg_insert(MediaObject)
        .values(
            digest=prepared.digest, extension=prepared.extension, data=prepared.data
        )
        .on_conflict_do_nothing(index_elements=[MediaObject.digest])
    )
    try:
        async with session.begin_nested():
            await session.execute(statement)
    except IntegrityError:
        # Only reachable if the row violates a CHECK — the digest shape or the
        # extension — which means this module and the migration disagree about
        # what it produces. Not a caller error, so it is not UnsupportedMedia.
        raise

    return prepared.stored()


async def load_image(
    digest: str, extension: str, *, session: AsyncSession
) -> MediaObject | None:
    """The row behind one media path, or None if we never stored it.

    ``extension`` is checked against the row rather than trusted, so the same
    bytes cannot be served under a second path with a different content type —
    the digest alone decides identity, and a mismatch means the URL was made up
    rather than issued by this store.
    """
    row = await session.get(MediaObject, digest)
    if row is None or row.extension != extension:
        return None
    return row


async def media_digests(session: AsyncSession) -> set[str]:
    """Every digest currently in the store. Used by the import script."""
    result = await session.execute(select(MediaObject.digest))
    return set(result.scalars())


def image_node(src: str, alt: str) -> dict[str, Any]:
    """The document node for a stored image, at full column width.

    Lives here rather than in ``tiptap.py`` for two reasons. This module owns
    ``MEDIA_SRC_RE``, ``MEDIA_ALIGNMENTS`` and the width bounds, so the node it
    builds is exactly the shape it validates a few lines below — the two cannot
    drift apart while they are in one file. And ``media.py`` imports
    ``tiptap.py``, so the dependency in the other direction would be a cycle.

    Used by the pipeline to place a generated lead image; a reviewer's uploads
    are built client-side by ``useMediaInsert.ts`` to the same shape. Both are
    then re-checked by ``assert_media_is_ours``, which is the point: nothing is
    trusted because of where it came from.
    """
    return {
        "type": "image",
        "attrs": {
            "src": src,
            "alt": alt,
            # Explicit rather than omitted. Absent is valid — every article
            # written before media could be laid out has neither attribute —
            # but a node this system writes should say what it means, and the
            # editor's controls read these two to render their initial state.
            "width": MEDIA_MAX_WIDTH,
            "align": "none",
        },
    }


def assert_media_is_ours(doc: dict[str, Any]) -> None:
    """Every media node points at something we control, laid out how we allow.

    Walks the whole tree, not just where media is meant to sit. A check that
    only looks in the expected place is not a check.

    Raises:
        UnsafeMediaError: naming the node and the offending value, because the
            reviewer needs to know which block to remove.
    """
    for node in iter_nodes(doc):
        node_type = node.get("type")
        if node_type not in ("image", "youtube"):
            continue

        attrs = node.get("attrs")
        attrs = attrs if isinstance(attrs, dict) else {}
        _assert_layout_is_allowed(attrs)

        if node_type == "image":
            src = attrs.get("src")
            if not isinstance(src, str) or not MEDIA_SRC_RE.match(src):
                raise UnsafeMediaError(
                    f"an image in this draft points at {_describe(src)}, which was "
                    "not uploaded through the console. Remove it and add the image "
                    "with the Image button — pasted or linked images are not stored, "
                    "so they would break or track readers."
                )
        else:
            video_id = attrs.get("videoId")
            if not isinstance(video_id, str) or not YOUTUBE_ID_RE.match(video_id):
                raise UnsafeMediaError(
                    f"a video block in this draft carries {_describe(video_id)} "
                    "instead of a YouTube video id. Remove it and add the video "
                    "with the Video button."
                )


def _assert_layout_is_allowed(attrs: dict[str, Any]) -> None:
    """``width`` and ``align`` are within the range the renderer can express.

    **Absent is valid.** Every article written before media could be laid out
    has neither attribute, and those documents must keep saving — a check that
    invalidates existing rows is a migration wearing a validator's clothes.

    Present-but-wrong is refused on the same argument as a bad ``src``: layout
    is set by a control, whole, never half-typed, so there is no mid-edit state
    to be caught in. What it keeps out is a ``width`` of 1e9 reaching a reader.
    """
    align = attrs.get("align")
    if align is not None and align not in MEDIA_ALIGNMENTS:
        raise UnsafeMediaError(
            f"a media block in this draft is aligned {_describe(align)}, which is "
            f"not one of {', '.join(sorted(MEDIA_ALIGNMENTS))}. Re-set the wrap "
            "with the controls on the block."
        )

    width = attrs.get("width")
    if width is None:
        return
    # bool is an int in Python, and `width: true` is not a width.
    if isinstance(width, bool) or not isinstance(width, int | float):
        raise UnsafeMediaError(
            f"a media block in this draft has a width of {_describe(width)}, which "
            "is not a number. Re-set the size with the controls on the block."
        )
    if not MEDIA_MIN_WIDTH <= width <= MEDIA_MAX_WIDTH:
        raise UnsafeMediaError(
            f"a media block in this draft is {width}% wide, outside the "
            f"{MEDIA_MIN_WIDTH}-{MEDIA_MAX_WIDTH}% a column can show. Re-set the "
            "size with the controls on the block."
        )


def _describe(value: object) -> str:
    """A short, quoted rendering of a rejected value for the error message."""
    if isinstance(value, str):
        return repr(value if len(value) <= 80 else value[:77] + "…")
    if value is None:
        return "nothing"
    return repr(value)
